pyk/src/pyk/kast/lexer.py

```diff
@@ -0,0 +1,202 @@
+from enum import Enum, auto
+from typing import TYPE_CHECKING, NamedTuple
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+    from typing import Final
+
+
+class TokenType(Enum):
+    EOF = auto()
+    LPAREN = auto()
+    RPAREN = auto()
+    COMMA = auto()
+    KSEQ = auto()
+    DOTK = auto()
+    DOTKLIST = auto()
+    TOKEN = auto()
+    ID = auto()
+    VARIABLE = auto()
+    KLABEL = auto()
+    STRING = auto()
+
+
+class Token(NamedTuple):
+    text: str
+    type: TokenType
+
+
+def lexer(text: Iterable[str]) -> Iterator[Token]:
+    it = iter(text)
+    la = next(it, '')
+    while True:
+        while la.isspace():
+            la = next(it, '')
+
+        if not la:
+            yield _TOKENS[TokenType.EOF]
+            return
+
+        try:
+            sublexer = _SUBLEXER[la]
+        except KeyError:
+            raise _unexpected_char(la) from None
+
+        token, la = sublexer(la, it)
+        yield token
+
+
+_TOKENS: Final = {
+    typ: Token(txt, typ)
+    for typ, txt in (
+        (TokenType.EOF, ''),
+        (TokenType.LPAREN, '('),
+        (TokenType.RPAREN, ')'),
+        (TokenType.COMMA, ','),
+        (TokenType.KSEQ, '~>'),
+        (TokenType.DOTK, '.K'),
+        (TokenType.DOTKLIST, '.KList'),
+        (TokenType.TOKEN, '#token'),
+    )
+}
+
+
+_DIGIT: Final = set('0123456789')
+_LOWER: Final = set('abcdefghijklmnopqrstuvwxyz')
+_UPPER: Final = set('ABCDEFGHIJKLMNOPQRSTUVWXYZ')
+
+
+_UNEXPECTED_EOF: Final = ValueError('Unexpected end of file')
+
+
+def _unexpected_char(actual: str, expected: str | None = None) -> ValueError:
+    if expected is None:
+        return ValueError(f'Unexpected character: {actual!r}')
+    actual_str = repr(actual) if actual else '<EOF>'
+    return ValueError(f'Expected {expected!r}, got: {actual_str}')
+
+
+SubLexer = Callable[[str, Iterator[str]], tuple[Token, str]]
+
+
+def _simple(token: Token) -> SubLexer:
+    def sublexer(la: str, it: Iterator[str]) -> tuple[Token, str]:
+        la = next(it, '')
+        return token, la
+
+    return sublexer
+
+
+def _delimited(delimiter: str, type: TokenType) -> SubLexer:
+    assert len(delimiter) == 1
+
+    def sublexer(la: str, it: Iterator[str]) -> tuple[Token, str]:
+        # assert la == delimiter
+        buf = [la]
+        la = next(it, '')
+        while True:
+            if not la:
+                raise _UNEXPECTED_EOF
+
+            elif la == delimiter:
+                buf.append(la)
+                la = next(it, '')
+                return Token(''.join(buf), type), la
+
+            elif la == '\\':
+                buf.append(la)
+                la = next(it, '')
+                if not la:
+                    raise _UNEXPECTED_EOF
+                buf.append(la)
+                la = next(it, '')
+
+            else:
+                buf.append(la)
+                la = next(it, '')
+
+    return sublexer
+
+
+def _kseq(la: str, it: Iterator[str]) -> tuple[Token, str]:
+    # assert la == '~'
```

Comment:
There are a few of these comments; should they be removed or reinstated?

Reply:
I reinstated them.